tests/test_synced_collection.py

```diff
@@ -0,0 +1,478 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+import pytest
+import uuid
+import os
+from tempfile import TemporaryDirectory
+from collections.abc import MutableMapping
+from collections.abc import MutableSequence
+
+from signac.core.collection_api import SyncedCollection
+from signac.core.jsoncollection import JSONDict
+from signac.core.jsoncollection import JSONList
+from signac.errors import InvalidKeyError
+from signac.errors import KeyTypeError
+
+
+FN_JSON = 'test.json'
+
+
+class TestSyncedCollectionBase():
+
+    _type = None
+    _write_concern = False
+
+    @pytest.fixture(autouse=True)
+    def setUp(self, request):
+        self._tmp_dir = TemporaryDirectory(prefix='jsondict_')
+        request.addfinalizer(self._tmp_dir.cleanup)
+        self._fn_ = os.path.join(self._tmp_dir.name, FN_JSON)
```

Comment:
Since this a completely new test implementation for this class, I might suggest to use a more pytest-onic approach to providing fixtures: 
 - https://docs.pytest.org/en/stable/fixture.html#fixtures-as-function-arguments
 - https://docs.pytest.org/en/stable/fixture.html#fixture-finalization-executing-teardown-code

Afaik, the use of `setUp` in this way is just a backwards-compatible approach to mixing the old unittest style with pytest.

Reply:
You could also consider is using [hypothesis](https://hypothesis.readthedocs.io/en/latest/) to write tests. The idea is that instead of encoding specific key values (for instance) in your test, you ask hypothesis to generate keys for you. You could use that in almost all the methods below, and it's very good at catching edge cases that you might otherwise miss (for example, empty strings, or strings with forbidden key types, or nested dict/list combinations...). If you don't want to try that here, we could add that to the to-do list for this project.

Reply:
Let's not overload this PR by introducing more complexity.

Reply:
Agreed, I'll add it to the list of open tasks.